### ISSUE TITLE
Updated hashing scheme

### ISSUE THREAD

**Issue by TheAlgorythm** [0x6470/poc_Bitwarden2Hashcat#1]
I am struggling since a week cracking my sisters password for her (with Hashcat) and I think I was able to track the problem down to a change in the hashing scheme in Bitwarden.
I tried to crack my own password, but even that wasn't possible, as the `keyHash` differs from the output of this PoC. Interestingly the output of this PoC is the same as of https://bitwarden.com/help/crypto.html . Therefore I opened this (https://github.com/bitwarden/help/issues/ #803) issue for the Bitwarden team.
Please could you at least verify the problem?

**Comment by 0x6470** [comment]
I had an extremly brief look at this and can confirm your observations.
You might want to contact Bitwarden Inc. directly (https://bitwarden.com/contact/) or ask in the User-to-User forum (https://community.bitwarden.com/c/support/6) in order to get a quicker response.
Though I can't guarantee it, I will try to take a deeper look at this in the next days.

**Comment by TheAlgorythm** [comment]
Thanks for the confirmation. I will further try to look into the code, but I am unfamiliar with the structure. I contacted Bitwarden directly.

**Comment by TheAlgorythm** [comment]
I think, that I've found the issue. There was a [PR](https://github.com/bitwarden/jslib/pull/404) which changed the final iteration count for local hashing to 2. The relevant change is [here](https://github.com/bitwarden/jslib/blob/e02e663ce1aed94d42db00dcdb2e42bdd625f0dc/common/src/services/crypto.service.ts#L488):
```ts
const iterations = hashPurpose === HashPurpose.LocalAuthorization ? 2 : 1;
```
But I wasn't able test it, as I don't have access to my main computer at the moment.

**Comment by 0x6470** [comment]
You're right!
By changing the one in line 12 to two, the PoC works again.
When it comes to Hashcat, I'm not sure whether to implement a new "version" of hash or a new token for this iteration count...
Could you get back to me, when Bitwarden anwered your inquiry?

**Comment by TheAlgorythm** [comment]
I think, the version should be updated. But that is a decision for the Hashcat team to make. Would you be able to update bitwarden2hashcat, which should be nearly nothing, later on? I will open a issue at hashcat and would give you the answer, once the Bitwarden team sent it.

**Comment by TheAlgorythm** [comment]
>Thank you for checking up on this! 😊
You can learn more on this here: https://community.bitwarden.com/t/what-do-keyhash-enckey-and-encprivatekey-mean-in-data-json/14210 and more on our encryption in general here: https://bitwarden.com/help/security/ and here: https://bitwarden.com/help/article/bitwarden-security-white-paper/
The 'keyHash' value that's present in the data.json file is not a hash of your master password with your email address as salt, and it would not match the 'Master Password Hash' value that you'd get on the site.

From Bitwarden

I mailed them, that this doesn't answer the question.

**Comment by 0x6470** [comment]
That surely didn't answer the question... let's see.
Changing the bitwarden2hashcat tool would technically be very easy, as the 1 after "$bitwaren$" in a Bitwarden Hash (`$bitwarden$1*100000*bm9yZXBseUBoYXNoY2F0Lm5ldA==*zAXL7noQxkIJG82vWuqyDsnoqnKAVU7gE/8IRI6BlMs=`) represents the version. Changing that to two, would be extremely easy.
However, there is one problem. Due to Bitwarden's poor documentation, I don't really know if the new hashing scheme is being user everywhere (which I assume, at least vor client-side hashing) or not. Additionally, there is no identifier in the data.json, which would indicate this.
I might add a flag to the tool, which would manually set the version.
I've actually been working on a new bitwarden2hashcat tool for a longer time, but never really got to it. This is why I will push the changes after completion (hopefully in the next weeks). 
In the meantime and as soon as hashcat implemented the new version, you can change the one in [line 159](https://github.com/0x6470/bitwarden2hashcat/blob/b6434cc8f4b2d746f29f05fabdd9142388eed653/bitwarden2hashcat.py#L159) to two.

**Comment by TheAlgorythm** [comment]
Yeah the documentation is really bad. The hashing scheme migration should be done automatically and therefore I would set v2 as the default. Further you could add an option to input the parameters via CLI arguments.

 A new version of bitwarden2hashcat is not urgent as I don't think that a new hashcat version will be released immediately. I changed the PoC already as I'm not so familiar with OpenCL.

**Comment by TheAlgorythm** [comment]
Bitwarden sent an email:
>Thank you for your patience! 😊
>I looked into this matter internally, and I wanted to follow up on your prior message and let you know that you were correct. We performed a few updates approximately 7 months ago, and now when you log into your account, your master password is hashed twice and saved inside the data.json, but it's hashed only once when you need to send it to the cloud server (https://vault.bitwarden.com/#/) for authentication. The value that is in the data.json is hashed twice to prevent an attacker from potentially taking that value from your data.json and passing it as-is in an authentication attempt with the cloud server. This is the reason that the value in the webpage you were using (https://bitwarden.com/help/crypto.html) was off by 1 hash iteration; It's because that webpage was not updated to reflect that change.
>I hope you find this clear and helpful, and I thank you in advance for your understanding. Please do not hesitate to let me know if you have any questions.

**Comment by TheAlgorythm** [comment]
The new format is: `$bitwarden$2*100000*2*bm9yZXBseUBoYXNoY2F0Lm5ldA==*CWCy4KZEEw1W92qB7xfLRNoJpepTMSyr7WJGZ0/Xr8c=`
See https://github.com/hashcat/hashcat/pull/3202
